internal/terraform/transform_query.go

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package terraform
+
+import (
+	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/dag"
+)
+
+type QueryTransformer struct {
+	// queryPlan is true when we are planning list resources.
+	queryPlan bool
+
+	// if validate is true, we are in validate mode and should not exclude any resources.
+	validate bool
+}
+
+func (t *QueryTransformer) Transform(g *Graph) error {
+	if !t.queryPlan {
+		// if we are not running a query-specific operation, we don't need to transform the graph
+		// as query-related files would not have been part of the parsed config.
+		return nil
+	}
+
+	if t.validate {
+		// if we are validating query files, we validate all resources
+		return nil
+	}
+
+	// a set to hold the resources that we want to keep and vertices along its path.
+	keep := dag.Set{}
+
+	for v := range dag.SelectSeq[GraphNodeConfigResource](g.VerticesSeq()) {
+		// we only get here if we are building a query plan, but not validating.
+		//
+		// By now, the graph already contains all resources from the config, including non-list resources.
+		// We start from each list resource node, look at its ancestors, and keep all vertices along its path except for non-list resources.
+		// The implication is that if there is any other node within the ancestors that depends on the remove non-list resource,
+		// its evaluation will be an error.
+		if v.ResourceAddr().Resource.Mode == addrs.ListResourceMode {
+			keep.Add(v)
+			deps := g.Ancestors(v)
+			for node := range deps {
+				if _, ok := node.(GraphNodeConfigResource); !ok {
```

Comment:
There can be other types of resource nodes in the graph here, just skipping these is not a complete guarantee that you can't pickup something unusual. Note that `GraphNodeConfigResource` encompasses all resource modes, so you could be removing things you need as well. 

In the rare case of a managed resource is captured as a dependency of a list resource (which would probably just be through a provider for now), I would say it's probably OK to just continue and see what happens. You should be at least a `-refresh-only` type mode, so there shouldn't be any changes to plan.

Reply:
> Note that GraphNodeConfigResource encompasses all resource modes, so you could be removing things you need as well.

Because the outer loop is going over the entire graph, and keeping all list resource nodes, we would have kept it, even if it wasn't added in this internal loop.

Reply:
> I would say it's probably OK to just continue and see what happens. You should be at least a -refresh-only type mode, so there shouldn't be any changes to plan.

We could also do that. But if it isn't a refresh mode, e.g. the resource is not yet in the state, the operation cannot succeed. But it would not succeed anyway if the resource is unknown

Reply:
Exactly! It can't work either way, so we probably don't need to go out of our way to avoid it, while also risking other inconsistencies in the graph.